activerecord/lib/active_record/connection_adapters/abstract_adapter.rb

```diff
@@ -740,6 +740,13 @@ def arel_visitor
 
         def build_statement_pool
         end
+
+        def metadata
```

Comment:
Those who wanted to leverage metadata support would patch or subclass from another adapter, for instance:

```ruby
class MyAdapter < ActiveRecord::ConnectionAdapters::Mysql2Adapter
  def metadata
    system_variables = @connection.session_track(Mysql2::Client::SESSION_TRACK_SYSTEM_VARIABLES)
    return unless system_variables
    JSON.parse(system_variables)
  end
end
```

Reply:
Should the abstract implementation default to returning an empty hash?

Reply:
Sounds like a good idea. Would you prefer that we freeze an empty hash and use that as a default to reduce allocations?

Reply:
That's what I'd do, but I'm not familiar enough with the design of this part of Rails to know wether it would be expected or not.

Also if the default is frozen, it would be preferable that the actual implementation is frozen as well.

Reply:
Updated 👍